src/main/java/org/broadinstitute/hellbender/tools/walkers/readorientation/CollectDataForReadOrientationFilter.java

```diff
@@ -0,0 +1,218 @@
+package org.broadinstitute.hellbender.tools.walkers.readorientation;
+
+import htsjdk.samtools.util.SequenceUtil;
+import org.broadinstitute.barclay.argparser.Argument;
+import org.broadinstitute.barclay.argparser.CommandLineProgramProperties;
+import org.broadinstitute.hellbender.cmdline.StandardArgumentDefinitions;
+import org.broadinstitute.hellbender.cmdline.programgroups.VariantProgramGroup;
+import org.broadinstitute.hellbender.engine.*;
+import org.broadinstitute.hellbender.engine.filters.ReadFilter;
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.utils.BaseUtils;
+import org.broadinstitute.hellbender.utils.MathUtils;
+import org.broadinstitute.hellbender.utils.Nucleotide;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.pileup.ReadPileup;
+import org.broadinstitute.hellbender.utils.read.ReadUtils;
+import org.broadinstitute.hellbender.tools.walkers.readorientation.AltSiteRecord.AltSiteRecordTableWriter;
+import org.broadinstitute.hellbender.tools.walkers.readorientation.RefSiteHistogram.RefSiteHistogramWriter;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Created by Takuto Sato on 7/26/17.
+ */
+
+@CommandLineProgramProperties(
+        summary = "Traverse and collect data from a bam for Mutect2 Read Orientation Filter",
+        oneLineSummary = "Data collection for Mutect2 Read Orientation Filter",
+        programGroup = VariantProgramGroup.class
+)
+
+public class CollectDataForReadOrientationFilter extends LocusWalker {
+    public static final String ALT_DATA_TABLE_SHORT_NAME = "alt_table";
+    public static final String ALT_DATA_TABLE_LONG_NAME = "alt_data_table";
+
+    public static final String REF_HISTOGRAM_TABLE_SHORT_NAME = "ref_table";
+    public static final String REF_HISTOGRAM_TABLE_LONG_NAME = "ref_histogram_table";
+
+
+    @Argument(fullName = "min-mq", shortName = "mq", doc = "exclude reads below this quality from pileup", optional = true)
+    static int MINIMUM_MEDIAN_MQ = 20;
+
+    @Argument(fullName = "min-bq", shortName = "bq", doc = "exclude bases below this quality from pileup", optional = true)
+    static int MINIMUM_BASE_QUALITY = 10;
+
+    @Argument(fullName = ALT_DATA_TABLE_LONG_NAME,
+            shortName = ALT_DATA_TABLE_SHORT_NAME,
+            doc = "a tab-separated output table of pileup data over alt sites")
+    static File altDataTable = null;
+
+    @Argument(fullName = REF_HISTOGRAM_TABLE_LONG_NAME,
+            shortName = REF_HISTOGRAM_TABLE_SHORT_NAME,
+            doc = "a tab-separated output histogram of depths over ref sites")
+    static File refHistogramTable = null;
+
+    // for computational efficiency, for each reference context, we build a depth histogram over ref sites
+    private static RefSiteHistogram[] refSiteHistograms = new RefSiteHistogram[64];
```

Comment:
done

Reply:
Actually I want to use an array if that's alright - it's the right data structure for the use case (with a known, finite number of keys and a pretty straightforward mapping to indices).  It does adds a bit more code, but array feels more satisfying and, over billions of ref sites for WGS, it might even show a small difference in performance

Reply:
Never mind. I see that a Map is more flexible. For instance if we want to expand to more than 3mer. Map it is.